internal/native_image/rules.bzl

```diff
@@ -159,6 +164,45 @@ def _graal_binary_implementation(ctx):
             **run_params
         )
 
+    # if we are building a shared library, prepare `CcInfo` for it, taking care to merge/forward
+    # our static dependencies.
+    cc_info = None
+    if ctx.attr.shared_library:
+        cc_toolchain = find_cpp_toolchain(ctx)
+        out_lib = binary
+        feature_configuration = cc_common.configure_features(
+            ctx = ctx,
+            cc_toolchain = cc_toolchain,
+            requested_features = ctx.features,
+            unsupported_features = ctx.disabled_features,
+        )
+        linker_input = cc_common.create_linker_input(
+            owner = ctx.label,
+            libraries = depset(direct = [
+                cc_common.create_library_to_link(
+                    actions = ctx.actions,
+                    feature_configuration = feature_configuration,
+                    cc_toolchain = cc_toolchain,
+                    dynamic_library = out_lib,
+                ),
+            ]),
+        )
+        compilation_context = cc_common.create_compilation_context()
+        linking_context = cc_common.create_linking_context(
+            linker_inputs = depset(direct = [linker_input]),
+        )
+        cc_info = cc_common.merge_cc_infos(cc_infos = [
+            CcInfo(
+              compilation_context = compilation_context,
+              linking_context = linking_context,
+            ),
+        ] + [
+            dep[CcInfo] for dep in ctx.attr.deps if CcInfo in dep]
```

Comment:
This would result in the `native_image` target behaving like a regular `cc_library`, but that won't be correct as it statically links its deps. Instead, I think that we should emit [`CcSharedLibraryInfo`](https://github.com/bazelbuild/bazel/blob/897edaa8ce4c0c4d570d3f279ee96d02604dc5fd/src/main/starlark/builtins_bzl/common/cc/cc_shared_library.bzl#L41) to get the correct linker interop with `cc_binary`.